compliance/collection/nodeinventorizer/cachedInventorizer.go

```diff
@@ -0,0 +1,163 @@
+package nodeinventorizer
+
+import (
+	"os"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	timestamp "github.com/gogo/protobuf/types"
+	"github.com/pkg/errors"
+	cmetrics "github.com/stackrox/rox/compliance/collection/metrics"
+	"github.com/stackrox/rox/generated/internalapi/sensor"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/env"
+)
+
+// InventoryScanOpts control behaviour of all node inventory related functions
+type InventoryScanOpts struct {
+	NodeName            string
+	Scanner             NodeInventorizer
+	InventoryCachePath  string
+	BackoffFilePath     string
+	BackoffWaitCallback func(time.Duration)
+}
+
+// TriggerNodeInventory creates a Node Inventory to send to Sensor.
+// To reduce strain on the Node, a linear backoff is checked before collecting an inventory.
+// Additionally, a cached inventory from an earlier invocation may be used instead of a full inventory run if it is fresh enough.
+// Note: This does not prevent strain in case of repeated pod recreation, as both mechanisms are based on an EmptyDir.
+func TriggerNodeInventory(opts *InventoryScanOpts) (*sensor.MsgFromCompliance, error) {
+	// check for existing backoff, wait for specified duration if needed, then persist the new backoff duration
+	initialBackoff := env.NodeInventoryInitialBackoff.DurationSetting()
+	currentBackoff, err := getCurrentBackoff(opts)
+	if err != nil {
+		return nil, err
+	}
+	if *currentBackoff > initialBackoff {
+		log.Warnf("Found existing backoff - last scan may have failed. Waiting %v seconds before running next inventory", currentBackoff)
+		opts.BackoffWaitCallback(*currentBackoff)
+	}
+	nextBackoff := calcNextBackoff(*currentBackoff)
+	bErr := writeBackoff(nextBackoff, opts.BackoffFilePath)
+	if bErr != nil {
+		log.Warnf("Error while persisting inventory backoff interval: %v", bErr)
+	}
+
+	// check whether a cached inventory exists that is recent enough to use
+	cachedInventory := loadCachedInventory(opts)
+	if cachedInventory != nil && isInventoryInCacheDuration(cachedInventory) {
+		log.Debugf("Using cached scan created at %v", cachedInventory.GetScanTime())
+		return createAndObserveMessage(cachedInventory), nil
+	}
+
+	// if no inventory exists, or it is too old, collect a fresh one and save it to the cache
+	newInventory, err := opts.Scanner.Scan(opts.NodeName)
+	if err != nil {
+		return nil, err
+	}
+
+	err = persistInventoryToCache(newInventory, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	removeBackoff(opts.BackoffFilePath) // Remove backoff file as late as possible
+	return createAndObserveMessage(newInventory), nil
```

Comment:
Nitpicks:

1. Something like `// Remove backoff file as late as possible` + `so that any failure does not lead to immediate re-run of scanning which may overload the node`.
2. You can create and observe message before removing backoff if removing backoff should really be as late as possible.

Reply:
Done

Reply:
What I meant by (2) was

```go
	result := createMessage(newInventory)
	removeBackoff(opts.BackoffFilePath)
	return result, nil
```

This is optional in the current form.

Reply:
The observation has moved into the caller, but I have also moved the backoff removal directly before the return statement to accomodate for this.
Thanks!